lib/promscrape/client.go

```diff
@@ -364,16 +365,23 @@ func doRequestWithPossibleRetry(hc *fasthttp.HostClient, req *fasthttp.Request,
 		} else if err != fasthttp.ErrConnectionClosed && !strings.Contains(err.Error(), "broken pipe") {
 			return err
 		}
-		// Retry request after exponentially increased sleep.
-		maxSleepTime := time.Until(deadline)
-		if sleepTime > maxSleepTime {
-			return fmt.Errorf("the server closes all the connection attempts: %w", err)
-		}
-		sleepTime += sleepTime
-		if sleepTime > maxSleepTime {
-			sleepTime = maxSleepTime
+
+		// Special handling for cases when server closes the connection.
```

Comment:
Is there a way to know if error was for closed connection or no?

Reply:
No, unfortunately connection closed by any reason returns just the same error.

Reply:
Then I suggest to simplify the retry logic like in example below:
```
 func doRequestWithPossibleRetry(hc *fasthttp.HostClient, req *fasthttp.Request, resp *fasthttp.Response, deadline time.Time) error {
-       sleepTime := time.Second
        scrapeRequests.Inc()
-       for {
+       backoffTime := time.Second
+       for i := 0; ; i++ {
                // Use DoDeadline instead of Do even if hc.ReadTimeout is already set in order to guarantee the given deadline
                // across multiple retries.
                err := hc.DoDeadline(req, resp, deadline)
@@ -366,12 +366,11 @@ func doRequestWithPossibleRetry(hc *fasthttp.HostClient, req *fasthttp.Request,
                }
                // Retry request after exponentially increased sleep.
                maxSleepTime := time.Until(deadline)
+               // the first retry happens without delay (i=0) intentionally,
+               // as a special case for handling broken or stale connections.
+               sleepTime := backoffTime * time.Duration(i)
                if sleepTime > maxSleepTime {
-                       return fmt.Errorf("the server closes all the connection attempts: %w", err)
-               }
-               sleepTime += sleepTime
-               if sleepTime > maxSleepTime {
-                       sleepTime = maxSleepTime
+                       return fmt.Errorf("the server exhausted all the connection attempts (%d): %w", i+1, err)
                }
                time.Sleep(sleepTime)
                scrapeRetries.Inc()

```
note: not tested

Reply:
Thanks for the suggestion!
Refactored code to follow this approach